### PR TITLE
creating speak streaming client and api

### PR DESCRIPTION
**Pull request by SandraRodgers**
## Summary by CodeRabbit

- **New Features**
  - Introduced a default callback handler for live transcription and text-to-speech API.
  - Added various response types and constants for text-to-speech streaming.
  - Implemented a new `MessageRouter` for handling different message types.

- **Enhancements**
  - Updated `SpeakOptions` to include `Callback` and `CallbackMethod`.
  - Improved error handling with new error constants for connection and validation issues.

- **Examples**
  - Added an interactive text-to-speech example using the Deepgram SDK.

These changes enhance the text-to-speech functionality, provide better error handling, and introduce examples for easier implementation.

### CODE REVIEW

File: pkg/api/speak-stream/v1/default.go
Comment (dvonthenen) [reply pkg/api/speak-stream/v1/default.go:185]: ignore these types of duplicate errors for this PR

File: pkg/client/speak/client_stream.go
Comment (dvonthenen) [reply pkg/client/speak/client_stream.go:279]: Ignore for now.

File: pkg/client/speak/client_stream.go
Comment (dvonthenen) [reply pkg/client/speak/client_stream.go:396]: Ignore the complexity for right now.

File: pkg/api/speak-stream/v1/router.go
Comment (dvonthenen) [reply pkg/api/speak-stream/v1/router.go:142]: I disagree. There is adequate logging and also returning the error code from the function.

File: pkg/api/speak-stream/v1/router.go
Comment (dvonthenen) [reply pkg/api/speak-stream/v1/router.go:151]: Then the return type is unique. An error will be printed in the logs (and console) due to the `klog.V(1)` output. Dismissing.

File: pkg/api/speak-stream/v1/default.go
Comment (dvonthenen) [reply pkg/api/speak-stream/v1/default.go:211]: this is just an example for a callback handler. not something that is actually used in production. this is used for demo purposes only.

File: pkg/api/speak-stream/v1/router.go
Comment (dvonthenen) [inline pkg/api/speak-stream/v1/router.go:30]: Can we revert this to `DEEPGRAM_DEBUG`? We don't need debugging differentiation between REST and WS. You are most likely going to use one of the other and if you do happen to use both, you probably want to see the debug for both.

File: pkg/client/speak/client_stream.go
Comment (dvonthenen) [reply pkg/client/speak/client_stream.go:115]: If this is copied from LiveClient, I am sure the edge cases are handled. Will do a compare to make sure though.

File: pkg/client/speak/client_stream.go
Comment (dvonthenen) [inline pkg/client/speak/client_stream.go:386]: this is where you need to differentiate between a Text vs Binary payload.
```
if msgType == websocket.TextMessage {
    // check if text or binary here
    err := c.router.Message(byMsg)
    ...
} else if msgType == websocket.binaryMessage {
    // TODO: call your `c.route.Binary()` function
} else {
    // log it
}
```

File: pkg/api/speak-stream/v1/router.go
Comment (dvonthenen) [inline pkg/api/speak-stream/v1/router.go:143]: Needs a function for the audio stream:
```
func (r *MessageRouter) Binary(byMsg []byte) error {
// TODO
}
```

File: pkg/client/speak/client_stream.go
Comment (dvonthenen) [inline pkg/client/speak/client_stream.go:657]: This can be deleted since this is specific to Transcription.

**Review by dvonthenen — CHANGES_REQUESTED** [review]
This is looking great. Got some feedback for you!

File: pkg/api/speak-stream/v1/interfaces/types.go
Comment (dvonthenen) [inline pkg/api/speak-stream/v1/interfaces/types.go:47]: This should be `FlushResponse` to match the label here:
https://github.com/deepgram/deepgram-go-sdk/pull/239/files#diff-e36b1ac7d75bbe05c7622c4e58a6a2c284fc55758034a4a28725cd2c8dbcea14R12

File: pkg/api/speak-stream/v1/interfaces/types.go
Comment (SandraRodgers) [reply pkg/api/speak-stream/v1/interfaces/types.go:53]: Should I create a new file `pkg/client/interfaces/interfaces.go` to put it in? Or should it go in just the `types-speak.go` file in that folder?

File: examples/speak/stream/http/main.go
Comment (dvonthenen) [reply examples/speak/stream/http/main.go:65]: this is just an example to help drive the implementation

File: examples/speak/stream/http/main.go
Comment (dvonthenen) [inline examples/speak/stream/http/main.go:110]: Let's delete this line for GA. We can do that now and don't need to wait.
```
cOptions := &interfaces.ClientOptions{}
```

File: examples/speak/stream/http/main.go
Comment (dvonthenen) [reply examples/speak/stream/http/main.go:127]: you will get an error emitted into the callback when something goes wrong

File: pkg/api/speak-stream/v1/default.go
Comment (dvonthenen) [reply pkg/api/speak-stream/v1/default.go:56]: This is only a demo example. Can't address this at a later time.

File: pkg/api/speak-stream/v1/default.go
Comment (dvonthenen) [reply pkg/api/speak-stream/v1/default.go:87]: This is only a demo example. Can't address this at a later time.

File: pkg/api/speak-stream/v1/default.go
Comment (dvonthenen) [reply pkg/api/speak-stream/v1/default.go:118]: This is only a demo example. Can't address this at a later time.

File: pkg/api/speak-stream/v1/default.go
Comment (dvonthenen) [reply pkg/api/speak-stream/v1/default.go:155]: This is only a demo example. Can't address this at a later time.

File: pkg/api/speak-stream/v1/default.go
Comment (dvonthenen) [reply pkg/api/speak-stream/v1/default.go:190]: This is only a demo example. Can't address this at a later time.

File: pkg/api/speak-stream/v1/interfaces/interfaces.go
Comment (dvonthenen) [inline pkg/api/speak-stream/v1/interfaces/interfaces.go:10]: I would maybe just have a comment dividing the WS Text vs Binary messages.

Something like:
```
# These are WS TextMessage that are used for flow control.
...

# These are WS BinaryMessage that are used to send audio data to the client
...
```

File: pkg/api/speak-stream/v1/router.go
Comment (dvonthenen) [reply pkg/api/speak-stream/v1/router.go:143]: Sweet!

File: pkg/api/version/constants.go
Comment (dvonthenen) [reply pkg/api/version/constants.go:41]: That's just not correct @coderabbit. @SandraRodgers has it right!

File: pkg/client/speak/client_stream.go
Comment (dvonthenen) [reply pkg/client/speak/client_stream.go:114]: Just need to change the Logging text here.

File: pkg/client/speak/client_stream.go
Comment (dvonthenen) [reply pkg/client/speak/client_stream.go:124]: We send connection issues via the callback.

File: pkg/client/speak/client_stream.go
Comment (dvonthenen) [reply pkg/client/speak/client_stream.go:277]: It's a complex problem that we will refactor at a later date. Here is an open issue for it:
https://github.com/deepgram/deepgram-go-sdk/issues/242

File: pkg/client/speak/client_stream.go
Comment (dvonthenen) [reply pkg/client/speak/client_stream.go:530]: This is fine!

File: pkg/client/speak/client_stream.go
Comment (dvonthenen) [inline pkg/client/speak/client_stream.go:543]: Can we add:
```
klog.V(6).Infof("StreamClient.Reset() LEAVE\n")
```

File: pkg/client/speak/client_stream.go
Comment (dvonthenen) [inline pkg/client/speak/client_stream.go:548]: Can we add:
```
klog.V(6).Infof("StreamClient.Reset() LEAVE\n")
```

File: pkg/client/speak/client_stream.go
Comment (dvonthenen) [inline pkg/client/speak/client_stream.go:551]: Then also a success message here:
```
klog.V(4).Infof("Reset Succeeded\n")
```

File: pkg/client/speak/constants.go
Comment (dvonthenen) [inline pkg/client/speak/constants.go:13]: Delete this `const` block. None of these are being used.

**Review by dvonthenen — CHANGES_REQUESTED** [review]
This is amazing! Great job here!

I just left some minor feedback and dismissed a bunch of coderabbit reviews that didn't add value.

Just a clarifying Q, for your `MyCallback`'s `Binary()`, did you test to make sure you could play the file?

Overall, this is EXCELLENT! 🤘

File: pkg/client/speak/constants.go
Comment (SandraRodgers) [reply pkg/client/speak/constants.go:13]: The defaultDelayBetweenRetry is being used here: https://github.com/deepgram/deepgram-go-sdk/blob/8f388b7bd0cd40c95c1ba4e99da39f5b368e65d0/pkg/client/speak/client_stream.go#L217

I've left that one for now, unless you think that logic in internalConnectWithCancel is unneeded

File: pkg/api/speak-stream/v1/interfaces/interfaces.go
Comment (SandraRodgers) [reply pkg/api/speak-stream/v1/interfaces/interfaces.go:10]: @dvonthenen Do you mean like this:

```go
type SpeakMessageCallback interface {
        // These are WS TextMessage that are used for flow control.
	Metadata(md *MetadataResponse) error
	Flushed(fl *FlushResponse) error
	Error(er *ErrorResponse) error
	Close(cr *CloseResponse) error
	Open(or *OpenResponse) error
         // These are WS BinaryMessage that are used to send audio data to the client
	Binary(byMsg []byte) error 
}
```

File: pkg/client/speak/constants.go
Comment (dvonthenen) [reply pkg/client/speak/constants.go:13]: Ooops! I missed that. It looks like all of them based on the golintci feedback

File: pkg/api/speak-stream/v1/interfaces/interfaces.go
Comment (dvonthenen) [reply pkg/api/speak-stream/v1/interfaces/interfaces.go:11]: 👍

File: examples/speak/stream/http/main.go
Comment (dvonthenen) [reply examples/speak/stream/http/main.go]: sweet!

File: pkg/client/speak/client_stream.go
Comment (dvonthenen) [reply pkg/client/speak/client_stream.go:385]: I think if you delete this:
https://github.com/deepgram/deepgram-go-sdk/pull/239/files#diff-ab3fe3fda9954ab4fcd7b7f364aea8f13ba807b4a19daa582a7927244fe67593R400-R402

It will remove this warning.

Comment (dvonthenen) [reply pkg/client/speak/client_stream.go:385]: Remove line 400-402
```
				if err != nil {
					klog.V(1).Infof("StreamClient.listen(): router.Message failed. Err: %v\n", err)
				}
```

Comment (dvonthenen) [reply pkg/client/speak/client_stream.go:385]: Gotcha!

File: pkg/client/speak/client_stream.go
Comment (dvonthenen) [reply pkg/client/speak/client_stream.go:115]: Will defer this for right now

File: pkg/client/speak/client_stream.go
Comment (dvonthenen) [reply pkg/client/speak/client_stream.go:277]: Part of the PR that was just merged addresses some of this. We'll defer this for right now

File: pkg/client/speak/client_stream.go
Comment (dvonthenen) [reply pkg/client/speak/client_stream.go:404]: Defer this for right now. The goal currently is stable versus refactored.

File: pkg/api/speak-stream/v1/default.go
Comment (dvonthenen) [reply pkg/api/speak-stream/v1/default.go:56]: Defer this for right now

**Comment by dvonthenen** [comment]
This is ready to go, but we need to merge this PR https://github.com/deepgram/deepgram-go-sdk/pull/245 and cut a release before merging this one.

**Review by dvonthenen — APPROVED** [review]

**Comment by dvonthenen** [comment]
Hold this for right now, but I think we are in the clear.

**Pull request merged**
